neqo-transport/src/streams.rs

```diff
@@ -0,0 +1,464 @@
+                if stream_id.is_send_only(self.role) {
+                    return Err(Error::StreamStateError);
+                }
+
+                if let (_, Some(rs)) = self.obtain_stream(stream_id)? {
+                    rs.send_flowc_update();
+                }
+            }
+            Frame::StreamsBlocked { .. } => {
+                stats.streams_blocked += 1;
+                // We send an update evry time we retire a stream. There is no need to
+                // trigger flow updates here.
+            }
+            _ => unreachable!("This is not a stream RecoveryToken"),
+        }
+        Ok(())
+    }
+
+    pub fn write_frames(
+        &mut self,
+        priority: TransmissionPriority,
+        builder: &mut PacketBuilder,
+        tokens: &mut Vec<RecoveryToken>,
+        stats: &mut FrameStats,
+    ) -> Res<()> {
+        if priority == TransmissionPriority::Important {
+            // Send `DATA_BLOCKED` as necessary.
+            self.flow_control_sender
+                .borrow_mut()
+                .write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            // Send `MAX_DATA` as necessary.
+            self.flow_control_receiver
+                .borrow_mut()
+                .write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.recv_streams.write_frames(builder, tokens, stats)?;
+
+            self.remote_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+            self.remote_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+        }
+
+        self.send_streams
+            .write_frames(priority, builder, tokens, stats)
+    }
+
+    pub fn lost(&mut self, token: &RecoveryToken) {
+        match token {
+            RecoveryToken::Stream(st) => self.send_streams.lost(&st),
+            RecoveryToken::ResetStream { stream_id } => self.send_streams.reset_lost(*stream_id),
+            RecoveryToken::StreamDataBlocked { stream_id, limit } => {
+                self.send_streams.blocked_lost(*stream_id, *limit)
+            }
+            RecoveryToken::MaxStreamData {
+                stream_id,
+                max_data,
+            } => {
+                if let Ok((_, Some(rs))) = self.obtain_stream(*stream_id) {
```

Comment:
Do we need to use obtain_stream() in this context?  Can't we use `self.recv.get_mut(id)` instead?  We're really only safeguarding against the stream being already dropped here, we don't need to call all the logic involving `self.send`.